mutex_flock.go

```diff
@@ -10,36 +10,120 @@ import (
 	"path/filepath"
 	"sync"
 	"syscall"
+	"time"
 
 	"github.com/juju/errors"
 )
 
-type mutex struct {
-	fd int
-	mu sync.Mutex
+func acquire(spec Spec, timeout <-chan time.Time) (Releaser, error) {
+	done := make(chan struct{})
+	defer close(done)
+	select {
+	case result := <-acquireFlock(spec.Name, done):
+		if result.err != nil {
+			return nil, errors.Trace(result.err)
+		}
+		return result.m, nil
+	case <-timeout:
+		return nil, ErrTimeout
+	case <-spec.Cancel:
+		return nil, ErrCancelled
+	}
 }
 
-func acquire(name string) (Releaser, error) {
-	flockName := filepath.Join(os.TempDir(), "juju-"+name)
-	fd, err := syscall.Open(flockName, syscall.O_CREAT|syscall.O_RDONLY, 0600)
-	if err != nil {
-		return nil, errors.Trace(err)
+var (
+	mu     sync.Mutex
+	active = make(map[string][]*waiter)
+)
+
+type waiter struct {
+	ch   chan<- acquireResult
+	done <-chan struct{}
+}
+
+type acquireResult struct {
+	m   Releaser
+	err error
+}
+
+// acquireFlock returns a channel on which an acquireResult will be
+// sent when the mutex acquisition completes, successfully or not.
+//
+// In Go, there is no natural way to cancel a flock syscall, so we
+// allow a flock syscall to continue on even when there is nothing
+// waiting for it, and then release the lock in that case. To avoid
+// an unbounded collection of goroutines, we ensure that there is
+// only one goroutine making a flock syscall at a time, per mutex
+// name.
+func acquireFlock(name string, done <-chan struct{}) <-chan acquireResult {
+	result := make(chan acquireResult)
+	w := &waiter{result, done}
+
+	mu.Lock()
+	defer mu.Unlock()
+	if waiters, ok := active[name]; ok {
+		active[name] = append(waiters, w)
+		return result
 	}
 
-	err = syscall.Flock(fd, syscall.LOCK_EX|syscall.LOCK_NB)
-	if err != nil {
-		syscall.Close(fd)
+	flockName := filepath.Join(os.TempDir(), "juju-"+name)
+	flock := func() (Releaser, error) {
+		fd, err := syscall.Open(flockName, syscall.O_CREAT|syscall.O_RDONLY, 0600)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if err := syscall.Flock(fd, syscall.LOCK_EX); err != nil {
+			syscall.Close(fd)
+			return nil, errors.Trace(err)
+		}
+		return &mutex{fd: fd}, nil
+	}
+	acquire := func() bool {
+		releaser, err := flock()
+		result := acquireResult{releaser, err}
+		mu.Lock()
+		defer mu.Unlock()
 
-		if err == syscall.EWOULDBLOCK {
-			return nil, errLocked
+		var sent bool
+		waiters := active[name]
+		for !sent && len(waiters) > 0 {
+			w, waiters = waiters[0], waiters[1:]
+			select {
+			case w.ch <- result:
+				sent = true
+			case <-w.done:
+				// waiter is done, so just
+				// remove it from the list
+				// and try the next one
+			}
+		}
```

Comment:
This seems odd that we grab a single flock() and then run all waiters.
Does that mean if we have 10 goroutines that want the lock in this process, and 10 in another process, all of this process will run if we get the lock, before we release the flock and let any of their 10 run?

I think it would be better to drop and re-acquire the flock for every waiter.

Reply:
Note the "!sent" in the for condition: we stop as soon as we've sent to one of them.